### PR TITLE
Add local chat echo support

### PR DESCRIPTION
**Pull request by peppy**
Another attempt (replaces #1124).

### CODE REVIEW

File: osu.Game/Online/Chat/Channel.cs [hunk 1]
```diff
@@ -52,11 +65,42 @@ public void AddNewMessages(params Message[] messages)
...
 
+        /// <summary>
+        /// Replace or remove a message from the channel.
+        /// </summary>
+        /// <param name="echo">The local echo message (client-side).</param>
+        /// <param name="final">The response message, or null if the message became invalid.</param>
+        public void ReplaceMessage(LocalEchoMessage echo, Message final)
+        {
+            if (!pendingMessages.Remove(echo))
+                Trace.Fail("Attempted to remove echo that wasn't present");
```
Comment (smoogipooo) [inline osu.Game/Online/Chat/Channel.cs]: This should be an exception, not a trace. It is a possible route for a caller given that this is a public function.

File: osu.Game/Online/Chat/Message.cs [hunk 2]
```diff
@@ -37,14 +37,23 @@ public Message()
...
-        public Message(long id)
+        public Message(long? id)
         {
             Id = id;
         }
 
-        public int CompareTo(Message other) => Id.CompareTo(other.Id);
+        public int CompareTo(Message other)
+        {
+
```
Comment (smoogipooo) [inline osu.Game/Online/Chat/Message.cs]: Remove space.

File: osu.Game/Overlays/Chat/DrawableChannel.cs [hunk 3]
```diff
@@ -90,6 +105,24 @@ private void newMessagesArrived(IEnumerable<Message> newMessages)
...
+        private void pendingMessageResolved(Message existing, Message updated)
+        {
+            var found = flow.Children.LastOrDefault(c => c.Message == existing);
+            if (found != null)
+            {
+                Trace.Assert(updated.Id.HasValue, "An updated message was returned with no ID.");
+
+                flow.Remove(found);
+                found.Message = updated;
+                flow.Add(found);
```
Comment (smoogipooo) [inline osu.Game/Overlays/Chat/DrawableChannel.cs]: Is it still necessary to remove and re-add like this? As far as I can see this doesn't do any depth update or the like...

**Review by smoogipoo — CHANGES_REQUESTED** [review]
This looks pretty good in general, nice structure

File: osu.Game/Overlays/Chat/DrawableChannel.cs (hunk 3 above)
Comment (peppy) [inline osu.Game/Overlays/Chat/DrawableChannel.cs]: yes

**Review by smoogipoo — DISMISSED** [review]

**Review by smoogipoo — APPROVED** [review]

**Pull request merged**